docs/_docs/development/threading.md

```diff
@@ -4,4 +4,357 @@ layout: docs
+```
+dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+    printf("block 1\n");
+});
+
+dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+    printf("block 2\n");
+});
+
+dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+    printf("block 3\n");
+});
+
+dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+    printf("block 4\n");
+});
+```
+produces the following output
+```
+block 1
+block 2
+block 3
+block 4
+```
+Starting the debugger at the end of the function should show 4 threads, each with the same queue attached. Now GCD is federating your work onto threads on demand. How many threads will it create?
+
+```
+for (int i=0; i < 100; i++) {
+    dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+        printf("block %i\n", i);
+    });
+}
+```
+Starting the debugger approximately 500ms after the completion of the for loop shows that there are 64 threads in the thread pool. The behavior you would experience by the thread pool limit is a delay in block execution after the 64th long-running block has been queued. If you are performing a _lot_ of work, make sure to grab the correct priority queue  `dispatch_get_global_queue(QOS_CLASS_xxxxxxx, 0)` for your needs. Otherwise you may denial-of-service any other blocks looking to be executed later on.
+
+An interesting dive into how cores fit into this picture can be found on Mike Ash's [Observing the A11 heterogenous cores](https://www.mikeash.com/pyblog/friday-qa-2017-11-10-observing-the-a11s-heterogenous-cores.html)
+
+Here we can do a fun test to understand the recycling behavior of GCD
+```
+const int64_t kOneMillion = 1000 * 1000;
+static mach_timebase_info_data_t s_timebase_info;
+
+static dispatch_once_t onceToken;
+dispatch_once(&onceToken, ^{
+    (void) mach_timebase_info(&s_timebase_info);
+});
+
+for (int i=0; i < 1000; i++) {
+    uint64_t start = mach_absolute_time();
+    dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+        uint64_t end = mach_absolute_time();
+        uint64_t elapsed = end - start;
+        uint64_t nanosec = elapsed * s_timebase_info.numer / s_timebase_info.denom;
+        printf("%llu ns block %i\n", nanosec / kOneMillion, i);
+        usleep(10000);
+    });
+}
+```
+Looking at the logs, we can see how GCD reaches a maximum background thread count, varied by CPU. This example was run on the 8 core Macbook Pro
+```
+1 ns block 62
+1 ns block 63
+10 ns block 65
+10 ns block 66
+10 ns block 64
+```
+
+## ASMainSerialQueue
+
+The `ASMainSerialQueue` ensures that work can be performed serially on the main thread without being interrupted. The key difference between this and purely using `dispatch_async(dispatch_get_main_queue, block)` is that the main thread can be interrupted between execution of blocks in its queue, where as this interface will execute everything possible in its queue on the main thread before returning control back to the OS.
```

Comment:
Are you saying that OS will not context switch away from the main thread while a block is being run by ASMainSerialQueue? I'm not sure that's the case.

If I remember correctly, the purpose of ASMainSerialQueue is that a block submitted off the main thread to the main thread will be run before another block submitted on the main thread if it was submitted after. If we just used `ASPerformBlockOnMainThread` a block submitted on the main thread *after* a block submitted off the main thread would be run before the block submitted (temporally) before.

Reply:
yes thats what i was trying to get at with my explanation. here i made a sweet google drawing to hopefully communicate my understanding of this class
![asmainserialqueue](https://user-images.githubusercontent.com/4482633/51281971-cc9bd980-1998-11e9-83f6-9042a1506731.png)

Reply:
also I'll change that from `returning control back to the OS` to `returning control back to GCD scheduler`

Reply:
In a ridiculous, scenario, is it possible for C to never execute if background threads are hammering ASMainSerialQueue with new blocks at rate that is equal to or faster than the rate at which main thread is consuming those blocks?